### PR TITLE
DOC: Update roadmap a bit more

### PR DESCRIPTION
**Pull request by seberg**
Just some thoughts, I didn't include ufunc chaining, although I think it would also be valid to mention it (even if I doubt anyone will work on it).  All others are things that have been discussed multiple times at least, so I think they are fine to add, even if I don't know if anyone will work on it:  That is true for practically everything.

I also always play with the idea that labeled axes might not be all that bad to think about.  But it may need a quite a lot of dedication. (It might not be that bad, but its a pretty broad thing.)

### CODE REVIEW

File: doc/neps/roadmap.rst [hunk 1]
```diff
@@ -97,6 +97,9 @@ first, and that could potentially be upstreamed into NumPy later, include:
   ``latin1``)
 - A unit dtype
 
+We further plan to extend the new API around ufuncs as the needs arise.
+One possibility here is creating a new, more powerful, API to allow hooking
+into existing NumPy ufunc implementations.
```
Comment (seberg) [inline doc/neps/roadmap.rst:102]: Note sure anyone can understand it, but I don't mind :).  I am talking about `pnumpy` here, we should have a better story for it (or decide to not have a story at all).

File: doc/neps/roadmap.rst [hunk 2]
```diff
@@ -29,8 +29,6 @@ needed based on real-world experience and needs.
 
 In addition, the ``__array_ufunc__`` and ``__array_function__`` protocols
 fulfill a role here - they are stable and used by several downstream projects.
-They do not cover the whole API, so use of the array API standard is preferred
-for new code.
```
[the remark this answers is not in the document]
Comment (rgommers) [reply doc/neps/roadmap.rst:33]: Seems fine to me to delete.

File: doc/neps/roadmap.rst [hunk 3]
```diff
@@ -51,6 +49,8 @@ Other performance improvement ideas include:
 
 - A better story around parallel execution (related is support for free-threaded
   CPython, see further down).
+- Using recent improvements in our ufunc context to allow users for example to
```
Comment (rgommers) [inline doc/neps/roadmap.rst:52]: Can you rephrase this slightly, removing "our" and perhaps pointing the API that is meant here?

File: doc/neps/roadmap.rst (hunk 1 above)
Comment (rgommers) [reply doc/neps/roadmap.rst:102]: Change "new API around ufuncs" to "ufunc C API"?

File: doc/neps/roadmap.rst [hunk 4]
```diff
@@ -122,6 +125,13 @@ We intend to write a NEP covering the support levels we provide and what is
 required for a platform to move to a higher tier of support, similar to
 `PEP 11 <https://peps.python.org/pep-0011/>`__.
 
+Further consistency fixes to promotion and scalar logic
+```````````````````````````````````````````````````````
+NumPy 2 fixed many issues around promotion especially with respect to scalars.
```
Comment (rgommers) [inline doc/neps/roadmap.rst:130]:
```suggestion
NumPy 2.0 fixed many issues around promotion especially with respect to scalars.
```

File: doc/neps/roadmap.rst [hunk 5]
```diff
@@ -122,6 +125,13 @@ We intend to write a NEP covering the support levels we provide and what is
...
+We plan to continue slowly fixing remaining inconsistencies.
```
Comment (rgommers) [inline doc/neps/roadmap.rst:131]:
```suggestion
We plan to continue fixing remaining inconsistencies.
```

File: doc/neps/roadmap.rst [hunk 6]
```diff
@@ -122,6 +125,13 @@ We intend to write a NEP covering the support levels we provide and what is
...
+For example, NumPy converts 0-D objects to scalars and some promotions
```
Comment (rgommers) [inline doc/neps/roadmap.rst:132]:
```suggestion
For example, NumPy converts 0-D objects to scalars, and some promotions
```

Comment (rgommers) [reply doc/neps/roadmap.rst:132]: Do we (you, or anyone else) actually plan to do anything about 0-D arrays vs. scalars? If not, we shouldn't mention it here.

**Review by rgommers — COMMENTED** [review]
Thanks Sebastian. Overall LGTM, a few comments.

File: doc/neps/roadmap.rst (hunk 1 above)
Comment (seberg) [reply doc/neps/roadmap.rst:102]: Thanks, better.

File: doc/neps/roadmap.rst (hunk 3 above)
Comment (seberg) [reply doc/neps/roadmap.rst]: I added ``errstate`` in brackets, which is what I mean (mostly).  Since I think that is a path towards switching ufunc loops, but of course I also mean the new C API which allows us to actually propagate that information.

File: doc/neps/roadmap.rst (hunk 6 above)
Comment (seberg) [reply doc/neps/roadmap.rst]: Not promising to do this right now, but that is true for most points on the list.

So yes, I think this is much more concrete than most points.  In fact, it is even mostly done, since I did most (maybe all) of the foundational work in NumPy 2!  (and one of that was another small compat break ;).)

File: doc/neps/roadmap.rst (hunk 1 above)
Comment (mhvk) [reply doc/neps/roadmap.rst:105]: I had never heard of `pnumpy`, but it seems interesting. Maybe add at end "... implementations, e.g., to allow parallel execution as done by [``pnumpy``](https://quansight.github.io/pnumpy/stable/index.html)"

**Review by mhvk — APPROVED** [review]
Nice. One suggestion in-line, but fine without too.

File: doc/neps/roadmap.rst (hunk 1 above)
Comment (seberg) [reply doc/neps/roadmap.rst:105]: Not sure how maintained it still is (or quite how well it worked in practice), but if others think it is stable enough to link, sure.
(Also, parallel execution belongs on a different level IMO, but that is part of the potential things here.)

Comment (rgommers) [reply doc/neps/roadmap.rst:105]: It's not maintained. It was an experiment, it worked and yielded interesting performance gains, but it wasn't interesting enough to pursue further at that point. I think it is an example of the types of things people would like to be able to do though, so it can serve as an example use case I'd say. @mattip may have more thoughts on it.

Comment (mhvk) [reply doc/neps/roadmap.rst:105]: If it is not pursued anymore then linking to it is perhaps not such a good idea... But it helps to give some example of why linking into the implementations might be useful.

Comment (mattip) [reply doc/neps/roadmap.rst:105]: pnumpy is unmaintained. I wouldn't want to draw more attention to it.

**Comment by rootsmusic** [comment]
Link #26157 to [Support for free-threaded CPython](https://numpy.org/neps/roadmap.html#support-for-free-threaded-cpython)?  (Since there's only one tracking issue, I assume that the roadmap isn't to release incremental upgrades toward full compatibility with Python 3.13.)

**Comment by rootsmusic** [comment]
See issue #23059

**Comment by mattip** [comment]
Let's put this in, we can continue to iterate on it in future PRs.

**Comment by mattip** [comment]
Thanks @seberg and reviewers
